plutus-core/plutus-core/src/Universe/Core.hs

```diff
@@ -0,0 +1,756 @@
+{-# LANGUAGE ConstraintKinds          #-}
+{-# LANGUAGE DataKinds                #-}
+{-# LANGUAGE FlexibleInstances        #-}
+{-# LANGUAGE GADTs                    #-}
+{-# LANGUAGE MultiParamTypeClasses    #-}
+{-# LANGUAGE PolyKinds                #-}
+{-# LANGUAGE QuantifiedConstraints    #-}
+{-# LANGUAGE RankNTypes               #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TemplateHaskell          #-}
+{-# LANGUAGE TypeApplications         #-}
+{-# LANGUAGE TypeFamilies             #-}
+{-# LANGUAGE TypeOperators            #-}
+{-# LANGUAGE UndecidableInstances     #-}
+-- Required only by 'Permits0' for some reason.
+{-# LANGUAGE UndecidableSuperClasses  #-}
+
+module Universe.Core
```

Comment:
Maybe we should split this a little? Like the encoding/decoding bit could be separate, except it's tied into `Closed` and I'm not sure why.

Reply:
So what makes a universe closed is that all of its type formers are fixed and no new ones can be added. E.g. quoting [this](https://golem.ph.utexas.edu/category/2011/04/homotopy_type_theory_v.html) (cut a little):

> the distinction as between an ‘open’ universe (to which new formers could be added) and a ‘closed’ universe (for which all the formers have been listed)

`Everywhere` is one way to enforce closedness and having a function that attempts to parse any sequence of integer tags as an encoded type tag is another, because that function is not extensible and so has to cover all possible cases and that is only possible in a closed universe. Plus, if you look at

```haskell
instance Closed uni => Hashable (SomeTypeIn uni) where
    hashWithSalt salt (SomeTypeIn uni) = hashWithSalt salt $ encodeUni uni

instance (Closed uni, uni `Everywhere` Hashable) => Hashable (ValueOf uni a) where
    hashWithSalt salt (ValueOf uni x) =
        bring (Proxy @Hashable) uni $ hashWithSalt salt (SomeTypeIn uni, x)
```

you can see that the constraints follow the same pattern as with `Eq`, `Show` etc, but we encode type tags to hash them in the `SomeTypeIn` instance. I.e. that notion of type tags being expressible as sequences of integer tags seems very interconnected with the notion of closedness under any constraints.

I could separate encoding/decoding into a separate class, but I don't see what that would gain us, apart from the opportunity to split up the module, half/most of which is literally documentation anyway. If anything, I'd separate instances and their docs (which amounts to a bit fewer than 200 lines). In some commits I did separate things into two modules, but then I decided that 750 lines is not a big deal, given that half/most of it is comments and merged them back.

Reply:
I get that it's not *so* much code, but it's quite a lot of conceptual weight, partly because of the docs! And there's all sorts of bits and pieces mixed in. I don't have a really good suggestion, though.

> Everywhere is one way to enforce closedness and having a function that attempts to parse any sequence of integer tags as an encoded type tag is another, because that function is not extensible and so has to cover all possible cases and that is only possible in a closed universe.

That seems to imply `Encodable => Closed`, but I'm not sure about the other direction. I don't *think* you can implement the encoding stuff with just `Everywhere`. It seems like there is maybe a refinement of something like `Constrainable = (has Everywhere) <= Partitionable (= tags can be disjointly assigned)`. Anyway, it's fine if this is more of a hassle than it's worth, I'm just trying to get my head around it.

Reply:
I'll try it out.

Reply:
> That seems to imply `Encodable => Closed`, but I'm not sure about the other direction.

I'm taking about semantics. If there's `Everywhere`, the universe is closed, using this definition of "closed":

> the distinction as between an ‘open’ universe (to which new formers could be added) and a ‘closed’ universe (for which all the formers have been listed)

If every type can be encoded/decoded, the universe is also closed in the same sense. It's not that `Closed` implies `Encodable` or something, it's just simply that open universes don't have these properties (how do you cover all the types in a certain instance/function at once if you don't know upfront what all these types are?).

I've expanded the comment a little to

> -- | A universe is 'Closed', if it's known how to constrain every type from the universe and
> -- every type can be encoded to / decoded from a sequence of integer tags.

So I really don't see much point in introducing a separate type class (and I can't come up with a good name for it anyway, it should be something universe-specific like `Closed` and related to encoding/decoding).